### PR TITLE
ReferencedLibraries

### PR DESCRIPTION
**Pull request by Purva-Chaudhari**
Pass adds referneced libraries to `linearizedBaseList`

### CODE REVIEW

File: example_contracts/libraries/using_for_simple.sol [hunk 1]
```diff
@@ -21,3 +13,12 @@ contract WARP {
     return t.amIReal(d);
   }
 }
+
+library Nums {
```
Comment (DomHenderson) [inline example_contracts/libraries/using_for_simple.sol:17]: What's the reasoning behind moving this?

File: src/passes/referencedLibraries.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import { ContractDefinition, ContractKind, FunctionCall, MemberAccess } from 'solc-typed-ast';
+import { AST } from '../ast/ast';
+import { ASTMapper } from '../ast/mapper';
+
+export class ReferencedLibraries extends ASTMapper {
+  visitFunctionCall(node: FunctionCall, ast: AST): void {
+    if (node.vExpression instanceof MemberAccess) {
+      const contractDef = new Map();
+
+      node.vExpression.context?.map.forEach((astNode, key) => {
```
Comment (DomHenderson) [inline src/passes/referencedLibraries.ts:11]: It's best to use `ast.context` in situations where a context is needed to remove the possibility of it being undefined

File: src/passes/referencedLibraries.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import assert = require('assert');
+import { ContractDefinition, ContractKind, FunctionCall, MemberAccess } from 'solc-typed-ast';
+import { AST } from '../ast/ast';
+import { ASTMapper } from '../ast/mapper';
+
+export class ReferencedLibraries extends ASTMapper {
+  visitFunctionCall(node: FunctionCall, ast: AST): void {
+    if (node.vExpression instanceof MemberAccess) {
+      const contractDef = new Map();
```
Comment (DomHenderson) [inline src/passes/referencedLibraries.ts:9]: ContractDef should have a type annotation to specify the key and value types of the map. This would improve readability and remove the need for asserting the type of the node later on

File: src/passes/referencedLibraries.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+      node.vExpression.context?.map.forEach((astNode, key) => {
+        if (astNode instanceof ContractDefinition && astNode.kind === ContractKind.Library) {
+          contractDef.set(key, astNode);
+        }
+      });
+
+      const func_id = node.vExpression.referencedDeclaration;
+
+      contractDef.forEach((astNode, key) => {
```
Comment (DomHenderson) [inline src/passes/referencedLibraries.ts:19]: This is clever, and as far as I can tell a completely valid way of doing this, but a comment block explaining how the function works would aid maintainability I think

File: src/passes/referencedLibraries.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        assert(astNode instanceof ContractDefinition);
+        if (astNode.vFunctions.some((func) => func.id === func_id)) {
+          const parent = node.getClosestParentByType(ContractDefinition);
+
+          if (parent === undefined) return;
```
Comment (DomHenderson) [inline src/passes/referencedLibraries.ts:24]: This brings up an interesting point, I was mistaken in believeing libraries could not be called from free functions. Handling that will be a separate PR I think, for now this should throw a NotSupportedYet exception if the library function call is not in a contract

**Review by DomHenderson — CHANGES_REQUESTED** [review]
This is looking good, just a few small changes needed. Also could you add some behaviour tests for libraries? I believe this pass was the last thing required for them to be used

File: example_contracts/libraries/using_for_simple.sol (hunk 1 above)
Comment (Purva-Chaudhari) [reply example_contracts/libraries/using_for_simple.sol:17]: Was using `visitContractDefinition` prior for a part to accomplish it. It was a check for library declared after contract. Can we keep it this was for behaviour testing?

Comment (DomHenderson) [reply example_contracts/libraries/using_for_simple.sol:17]: I'd suggest having one behaviour test with it declared before, one with it declared after, and if it works or is easily fixable, one with it being imported from an adjacent file. Might as well cover all bases, especially as the sourceUnitSplitter is liable to change in the future

**Review by DomHenderson — APPROVED** [review]
